src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -99,33 +184,31 @@ internal static string IPv6AddressToString(ushort[] address, uint scopeId)
                 buffer.Append('%').Append(scopeId);
             }
 
-            return StringBuilderCache.GetStringAndRelease(buffer);
+            return buffer;
         }
 
         private static unsafe void FormatIPv4AddressNumber(int number, char* addressString, ref int offset)
         {
-            int i = offset;
+            offset += number > 99 ? 3 : number > 9 ? 2 : 1;
+
+            int i = 0;
             do
             {
                 int rem;
                 number = Math.DivRem(number, 10, out rem);
-                addressString[--i] = (char)('0' + rem);
+                addressString[offset - ++i] = (char)('0' + rem);
```

Comment:
It just inhibits certain optimizations, as the compiler doesn't necessarily know whether the target is changing between uses.